_posts/contests/2021-02-05-Performance-analysis-and-tuning-contest-4.md

```diff
@@ -312,6 +312,237 @@ One contestant noticed that it is convenient to keep the arrays `delta_x` and `d
 
 **non_max_supp**
 
+The function `non_max_supp` was the most interesting one, at least according to me. Click here TODO(Add link) to see the full source code of the critical loop.
+
+The loop iterates over pixels row-wise. Inside the loop, for each pixel, the code performs the same operations. But the input values for the operation depend on the complex derivative values calculated in the previous step. Depending on the value of the derivatives, the code will perform the action on one of the eight neighboring pixels. Here is the full source code of the critical loop:
+
+```cpp
+for(rowcount=1,magrowptr=mag+ncols+1,gxrowptr=gradx+ncols+1,
+  gyrowptr=grady+ncols+1,resultrowptr=result+ncols+1;
+  rowcount<nrows-2; 
+  rowcount++,magrowptr+=ncols,gyrowptr+=ncols,gxrowptr+=ncols,
+  resultrowptr+=ncols){   
+  for(colcount=1,magptr=magrowptr,gxptr=gxrowptr,gyptr=gyrowptr,
+      resultptr=resultrowptr;colcount<ncols-2; 
+      colcount++,magptr++,gxptr++,gyptr++,resultptr++){   
+      m00 = *magptr;
+      if(m00 == 0){
+        *resultptr = (unsigned char) NOEDGE;
+      }
+      else{
+        xperp = -(gx = *gxptr)/((float)m00);
+        yperp = (gy = *gyptr)/((float)m00);
+      }
+
+      if(gx >= 0){
+        if(gy >= 0){
+                if (gx >= gy)
+                {  
+                    /* 111 */
```

Comment:
I suggest to get rid of all the code that is not relevant here. I'm afraid it's not readable in the current form anyway. Since you're not talking about specific computations, maybe we can go along with something like:
```
for (..)
  for (..)
      ...
      if(gx >= 0){
        if(gy >= 0){
          if (gx >= gy) {
            ...
          } else {
            ...
          }
```
I.e. just show the structure and avoid the details. Again, we can put a link to the baseline code.